test/distributed/test_c10d_nccl.py

```diff
@@ -2454,6 +2454,199 @@ def test_channels_last_contig(self):
         process_group.broadcast([tensor]).wait()
 
 
+# These are internal OpType enum values in Work.hpp, the value is subject to
+# change in future versions.
+BROADCAST = 0
+ALLREDUCE = 1
+ALLGATHER = 4
+
+
+class WorkHookTest(MultiProcessTestCase):
+
+    @property
+    def world_size(self):
+        return 2
+
+    def setUp(self):
+        super().setUp()
+        # set NCCL_ENABLE_TIMING to enable timing for CUDAEvents
+        # in ProcessGroup Work
+        os.environ["NCCL_ENABLE_TIMING"] = "1"
+        self._spawn_processes()
+
+    def tearDown(self):
+        super().tearDown()
+        try:
+            os.remove(self.file_name)
+        except OSError:
+            pass
+
+    def _get_store(self):
+        return dist.FileStore(self.file_name, self.world_size)
+
+    def _get_process_group(self):
+        store = self._get_store()
+        c10d.init_process_group("nccl", store=store, rank=self.rank, world_size=self.world_size)
+        return c10d.distributed_c10d._get_default_group()
+
+    @requires_nccl()
+    @skip_if_lt_x_gpu(2)
+    def test_on_completion_hook_broadcast(self):
+        pg = self._get_process_group()
+        num_hook_fired = 0
+        durations: List[float] = []
+
+        def hook(work):
```

Comment:
does work obj contain metadata about the op, the input sizes, the PG/other ranks?  i'm guessing yes but just to be sure, seems like all of that would be useful to have access to from the hook

Reply:
Yep, `WorkNCCL` has all of those except an direct access to PG, but we can add that if necessary either as a reference in Work or as an argument to the hook.